<robot name="tf_simulation3" xmlns:xacro="http://ros.org/wiki/xacro">
    <material name="black">
        <color rgba="0 0 0 0.7"/>
    </material>
    <material name="white">
        <color rgba="1 1 1 0.7"/>
    </material>

    <link name="base_link"/>
    <joint name="base_to_body_joint" type="fixed">
        <parent link="base_link"/>
        <child link="body_link"/>
    </joint>
    <link name="body_link">
        <visual>
            <geometry>
                <box size="0.2 0.3 0.1"/>
            </geometry>
            <origin rpy="0 0 0" xyz="0 0 0"/>
            <material name="white" />
        </visual>
        <inertial>
            <!-- <origin rpy="0 0 0" xyz="0 0 0"/> -->
            <mass value="0.1"/>
            <inertia ixx="0.001" ixy="0" ixz="0" iyy="0.001" iyz="0" izz="0.001"/>
        </inertial>
        <collision>
            <!-- <origin rpy="0 0 0" xyz="0 0 0"/> -->
            <geometry>
                <cylinder length="0.05" radius="0.05"/>
            </geometry>
        </collision>
    </link>
    <gazebo reference="body_link">
        <material>Gazebo/white</material>
    </gazebo>

    <joint name="body_to_ball_joint" type="fixed">
        <parent link="body_link"/>
        <child link="ball_link"/>
    </joint>
    <link name="ball_link">
        <visual>
            <origin rpy="0 0 0" xyz="0 0 0"/>
            <geometry>
                <sphere radius="0.025"/>
            </geometry>
            <material name="white" />
        </visual>
        <collision>
            <!-- <origin rpy="0 0 0" xyz="0 0 0"/> -->
            <geometry>
                <sphere radius="0.025"/>
            </geometry>
        </collision>
    </link>
    <gazebo reference="ball_link">
        <material>Gazebo/white</material>
        <mu1 value="0.5"/>
        <mu2 value="0.5"/>
    </gazebo>

    <joint name="base_to_wheel_R" type="fixed">
        <parent link="body_link"/>
        <child link="wheel_R"/>
    </joint>
    <link name="wheel_R">
        <visual>
            <geometry>
                <cylinder length="0.05" radius="0.05"/>
            </geometry>
            <origin rpy="0 0 0" xyz="0 0 0"/>
            <material name="black"/>
        </visual>
        <inertial>
            <!-- <origin rpy="0 0 0" xyz="0 0 0"/> -->
            <mass value="0.1"/>
            <inertia ixx="0.0001" ixy="0" ixz="0" iyy="0.0001" iyz="0" izz="0.0001"/>
        </inertial>
        <collision>
            <!-- <origin rpy="0 0 0" xyz="0 0 0"/> -->
            <geometry>
                <cylinder length="0.048" radius="0.05"/>
            </geometry>
        </collision>
    </link>
    <gazebo reference="wheel_R">
        <material>Gazebo/black</material>
        <mu1 value="1.0"/>
        <mu2 value="1.0"/>
        <kp value="50000"/>
        <kd value="10"/>
    </gazebo>

    <joint name="base_to_wheel_L" type="fixed">
        <parent link="body_link"/>
        <child link="wheel_L"/>
    </joint>
    <link name="wheel_L">
        <visual>
            <geometry>
                <cylinder length="0.05" radius="0.05"/>
            </geometry>
            <origin rpy="0 0 0" xyz="0 0 0"/>
            <material name="black"/>
        </visual>
        <inertial>
            <!-- <origin rpy="0 0 0" xyz="0 0 0"/> -->
            <mass value="0.1"/>
            <inertia ixx="0.0001" ixy="0" ixz="0" iyy="0.0001" iyz="0" izz="0.0001"/>
        </inertial>
        <collision>
            <!-- <origin rpy="0 0 0" xyz="0 0 0"/> -->
            <geometry>
                <cylinder length="0.048" radius="0.05"/>
            </geometry>
        </collision>
    </link>
    <gazebo reference="wheel_L">
        <material>Gazebo/black</material>
        <mu1 value="1.0"/>
        <mu2 value="1.0"/>
        <kp value="50000"/>
        <kd value="10"/>
    </gazebo>

    <joint name="laser_joint" type="fixed">
        <parent link="body_link"/>
        <child link="laser_link"/>
    </joint>
    <link name="laser_link">
        <visual>
            <origin rpy="0 0 0" xyz="0 0 0"/>
            <geometry>
                <cylinder length="0.04" radius="0.02"/>
            </geometry>
            <material name="rad">
                <color rgba="1 0 0 0.7"/>
            </material>
        </visual>
        <collision>
            <!-- <origin rpy="0 0 0" xyz="0 0 0"/> -->
            <geometry>
                <cylinder length="0.04" radius="0.02"/>
            </geometry>
        </collision>
    </link>
    <gazebo reference="laser_link">
        <material>Gazebo/Red</material>
        <sensor name="laser" type="ray">
        <pose>0 0 0 0 0 0</pose>
        <visualize>true</visualize>
        <update_rate>10</update_rate>
        <ray>
            <scan>
            <horizontal>
                <samples>400</samples>
                <resolution>1</resolution>
                <min_angle>-1.0471975512</min_angle>
                <max_angle>1.0471975512</max_angle>
            </horizontal>
            </scan>
            <range>
            <min>0.10</min>
            <max>20.0</max>
            <resolution>0.01</resolution>
            </range>
            <noise>
            <type>gaussian</type>
            <mean>0.0</mean>
            <stddev>0.01</stddev>
            </noise>
        </ray>
        <plugin filename="libgazebo_ros_laser.so" name="gazebo_ros_lidar_controller">
            <topicName>/laser_link/scan</topicName>
            <frameName>laser_link</frameName>
        </plugin>
        </sensor>
    </gazebo>

    <gazebo>
        <plugin filename="libgazebo_ros_control.so" name="gazebo_ros_control">
        <robotNamespace>/robot3_simulation</robotNamespace>
        </plugin>
    </gazebo>
</robot>

